src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts

```diff
@@ -122,58 +125,99 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
-                this.loadExistingStudentsData(existingStudentTable, this.existingStudents);
-              }
-              this.isExistingStudentsPresent = true;
-            });
-          },
-          error: (resp: ErrorMessageOutput) => {
-            this.enrollErrorMessage = resp.error.message;
-          },
+    );
+
+    const handledRequests: Observable<Students> = this.handleTimeoutForEnrollRequest(enrollRequests);
+    handledRequests.subscribe({
+      next: (resp: Students) => {
+        enrolledStudents.push(...resp.students);
+        this.remainingStudentChunks = this.remainingStudentChunks.slice(1, this.remainingStudentChunks.length);
+      },
+      complete: () => {
+        this.showEnrollResults = true;
+        this.statusMessage.pop(); // removes any existing error status message
+        this.statusMessageService.showSuccessToast('Enrollment successful. Summary given below.');
+
+        const allStudentEntries: StudentEnrollRequest[] = [];
+        this.allStudentChunks.forEach((chunk: StudentEnrollRequest[]) => allStudentEntries.push(...chunk));
+
+        this.enrollResultPanelList =
+            this.populateEnrollResultPanelList(this.existingStudents, enrolledStudents, allStudentEntries);
+
+        this.studentService.getStudentsFromCourse({ courseId: this.courseId }).subscribe((resp: Students) => {
+          this.existingStudents = resp.students;
+          if (!this.isExistingStudentsPanelCollapsed) {
+            const existingStudentTable: Handsontable = this.hotRegisterer.getInstance(this.existingStudentsHOT);
+            this.loadExistingStudentsData(existingStudentTable, this.existingStudents);
+          }
+          this.isExistingStudentsPresent = true;
         });
+      },
+      error: (resp: ErrorMessageOutput) => {
+        this.enrollErrorMessage = resp.error.message;
+      },
+    });
   }
 
+  private handleTimeoutForEnrollRequest(prevRequests: Observable<Students>): Observable<Students> {
+
+    return prevRequests.pipe(finalize(() => this.isEnrolling = false))
+        .pipe(takeWhile(() => this.remainingStudentChunks.length > 0))
+        .pipe(catchError(() => {
+          // First request in the list caused the timeout
+          const currentLongRequest: StudentEnrollRequest[] = this.remainingStudentChunks[0];
+
+          // Break the long request into half
+          const studentEnrollRequest1: StudentEnrollRequest[] =
+              currentLongRequest.slice(0, currentLongRequest.length / 2);
+
+          const studentEnrollRequest2: StudentEnrollRequest[] =
+              currentLongRequest.slice(currentLongRequest.length / 2, currentLongRequest.length);
+
+          this.remainingStudentChunks =
+              this.remainingStudentChunks.filter((chunk: StudentEnrollRequest[]) =>
+                  chunk !== currentLongRequest);
+          this.remainingStudentChunks =
+              [studentEnrollRequest1, studentEnrollRequest2, ...this.remainingStudentChunks];
```

Comment:
1. Why not just 
```
this.remainingStudentChunks = [
              studentEnrollRequest1, studentEnrollRequest2, 
              ...this.remainingStudentChunks.filter((chunk: StudentEnrollRequest[]) => chunk !== currentLongRequest);
];
```
2. Would it look more efficient (look-wise) to do this?
```
this.remainingStudentChunks.shift();
this.remainingStudentChunks.push(studentEnrollRequest1, studentEnrollRequest2);
```
honestly the computational cost is almost negligible, but doesn't look right to me to use a loop to pop head.

Reply:
I would like the new small requests to be at the beginning, so ```unshift()``` will be better?